### PR TITLE
Rectangular cell lists

### PR DESCRIPTION
**Pull request by kjs73**
This allows rectangular cell lists. Tested with LJ-cut potential function.

### CODE REVIEW

File: source/pele/cell_lists.h [hunk 1]
```diff
@@ -550,9 +550,13 @@ CellLists<distance_policy>::CellLists(
...
-      m_lattice_tool(dist, boxv, rcut, pele::Array<size_t>(m_ndim, 
-                  std::max<size_t>(1, (size_t)(ncellx_scale * boxv[0] / rcut))     //no of cells in one dimension
-                  )),
+      // http://stackoverflow.com/questions/13461538/lambda-of-a-lambda-the-function-is-not-captured
+      m_lattice_tool(dist, boxv, rcut,
+        [&ncellx_scale, &rcut](const pele::Array<double> boxv)
+        { pele::Array<size_t> res(boxv.size());
+        for (size_t i = 0; i < res.size(); ++i)
+        { res[i] = std::max<size_t>(1, ncellx_scale * boxv[i] / rcut) ; }
+        return res; }(boxv)),
```
Comment (js850) [inline source/pele/cell_lists.h]: for clarity, I would implement this lambda function as a static private member function or a function in the anonymous namespace

File: source/pele/cell_lists.h [hunk 2]
```diff
@@ -592,7 +592,7 @@ void CellLists<distance_policy>::setup(Array<double> coords)
 
     // print messages if any of the parameters seem bad
     size_t ncellx = m_lattice_tool.m_ncells_vec[0];
-    if (ncellx < 5) {
+    if (*std::min_element(m_lattice_tool.m_ncells_vec.data(), m_lattice_tool.m_ncells_vec.data() + m_ndim) < 5) {
```
Comment (js850) [inline source/pele/cell_lists.h]: I think you can use .begin() and .end() here.

File: source/pele/cell_lists.h [hunk 3]
```diff
@@ -603,7 +603,7 @@ void CellLists<distance_policy>::setup(Array<double> coords)
         std::cout << "CellLists: efficiency warning: the number of cells ("<<m_lattice_tool.m_ncells<<")"<<
                 " is greater than the number of atoms ("<<m_natoms<<").\n";
     }
-    if (m_lattice_tool.m_rcut > 0.5 * m_lattice_tool.m_boxvec[0]) {
+    if (m_lattice_tool.m_rcut > 0.5 * *std::min_element(m_lattice_tool.m_boxvec.data(), m_lattice_tool.m_boxvec.data() + m_ndim)) {
```
Comment (js850) [inline source/pele/cell_lists.h]: same as above

File: cpp_tests/source/test_cell_lists.cpp [hunk 4]
```diff
@@ -1167,3 +1169,70 @@ TEST(LatticeNeighborsTest, NonPeriodic_Works2)
...
+    for (size_t i = 0; i < ndim; ++i) {
+        boxvec[i] = std::pow((2 * rcut + 1), i + 1) + 0.8 * i;
+    }
+    const size_t N = 25;
+    pele::Array<double> x(ndim * N);
+    std::mt19937_64 gen(42);
+    for (size_t i = 0; i < N; ++i) {
+        x[i * ndim] = std::uniform_real_distribution<double>{0, boxvec[0]}(gen);
+        x[i * ndim + 1] = std::uniform_real_distribution<double>{0, boxvec[1]}(gen);
+    }
```
Comment (js850) [inline cpp_tests/source/test_cell_lists.cpp]: does it make sense to add another test where the atoms are not all in the [0,L] box.  e.g. in [-L/2, L/2] or scattered over all space.

**Comment by js850** [comment]
Good change, thanks.  Aside from my comments I think it looks good.  I'm surprised so little needed to be changed.

File: source/pele/cell_lists.h [hunk 5]
```diff
@@ -592,7 +592,7 @@ void CellLists<distance_policy>::setup(Array<double> coords)
 
     // print messages if any of the parameters seem bad
```
Comment (js850) [inline source/pele/cell_lists.h]: p.s. I was never happy with these efficiency warnings.  They didn't seem to work very well.  If you have a better idea of how to implement them feel free to shout out

**Comment by kjs73** [comment]
Thank you for the feedback! The points should be solved now. Concerning the warnings, I am not sure what to do instead either. We need to keep the warning about rcut, not sure about the others.

**Comment by js850** [comment]
:+1:

**Pull request merged**